docker-compose.yml

```diff
@@ -21,30 +21,25 @@ services:
     image: redis:4
     ports:
       - "6379:6379"
-    
+
   ocim:
     build:
       context: .
       dockerfile: Dockerfile.ocim
-    depends_on: 
+    depends_on:
       - postgresql
       - redis
     env_file: .env
-    environment: 
-      ALLOWED_HOSTS: '["*"]'
-      DATABASE_URL: '${DATABASE_URL:-postgres://opencraft@postgresql/opencraft}'
-      REDIS_URL: '${REDIS_URL:-redis://redis:6379}'
```

Comment:
The idea of having these variables here is that you don't have to set them in `.env`, especially since they are specific to the Docker setup.

`ALLOWED_HOSTS` is particularly problematic because, without the wildcard, you won't be able to access the Django app from outside the container.

Looking at the docs though, it would make sense to document that these don't need to be set.

Reply:
Thanks!  This is only a WIP dump of what I was working on btw; I'll refine it when I come back to it tomorrow :)

Reply:
@kewne so I didn't like this because there is no longer a single source of truth for the environment, and overriding it requires either overriding the vars somewhere else or directly editing the docker-compose.yml.  Also, I couldn't get it working with the default config there anyway.

Reply:
> I didn't like this because there is no longer a single source of truth for the environment

I agree with the sentiment; my reasoning is, once again, from the POV of someone trying to set this up for the first time: if the instructions tell me to set up a `.env` file with certain values, I'd expect them to be specific to my local environment (i.e. my OpenStack credentials, my MySQL credentials, etc); however, most of the time these three specific variables need to match the `docker-compose.yml` file and shouldn't be changed;
while you *can* change them, I think it's acceptable to do it differently from the other config because it's not the common use case.

The ideal thing here would really be to have a way to "stack" configuration (e.g. like devstack' `private.py` does), I just don't know how it could be achieved because neither Docker Compose or Honcho support it: the only thing I can think of is having a custom entrypoint script but that has other issues.

> I couldn't get it working with the default config there anyway

Can you describe what's not working? That would help me to better understand the issue and maybe find a better solution.

Reply:
> Can you describe what's not working? 

I had a lot of trouble setting the database_url variable.  It kept wanting to use the database_url set from the env file I reckon.  That's why I'm pushing for a single source of truth.

> from the POV of someone trying to set this up for the first time: if the instructions tell me to set up a .env file with certain values

I think this is fine - we're telling them in the docs exactly what variables are required for the docker devstack, and why.  This makes it clear, while also keeping the single source of truth and allowing flexibility to set whatever you want for different use cases.

Reply:
> I had a lot of trouble setting the database_url variable. It kept wanting to use the database_url set from the env file I reckon.

That shouldn't be possible, according to [the docker compose docs](https://docs.docker.com/compose/compose-file/compose-file-v3/#env_file):
> Environment variables declared in the environment section override these values – this holds true even if those values are empty or undefined.

I imagine it could be caused by having `DATABASE_URL` set in your environment though (which would use `.env` if  running something like `honcho run bash`).

Reply:
@kewne 

bq. That shouldn't be possible, according to the docker compose docs:

Ah right, the var was defined as:

```
'${DATABASE_URL:-postgres://opencraft@postgresql/opencraft}'
```

So it was never hitting the default, because DATABASE_URL was set to something else previously.

We _could_ remove the default fallback and hardcode the desired postgres url, but then it's impossible for the user to override it if they wish without editing the docker-compose file.